### PR TITLE
[MLIR] Add DIExpression to LLVM dialect

### PR DESCRIPTION
**Pull request by zyx-billy**
Add initial support for DIExpression in LLVM dialect.

Similar to LLVM IR, DI Expression is encoded as a list of uint64. The difference is that LLVM IR has helpers for understanding the expression (e.g. for verification and pretty printing), whereas the current support added by this PR treats the expression elements as opaque.

### CODE REVIEW

File: mlir/test/Dialect/LLVMIR/mem2reg-dbginfo.mlir [hunk 1]
```diff
@@ -36,17 +36,17 @@ llvm.func @block_argument_value(%arg0: i64, %arg1: i1) -> i64 {
   // CHECK-NOT: = llvm.alloca
   %1 = llvm.alloca %0 x i64 {alignment = 8 : i64} : (i32) -> !llvm.ptr
   // CHECK-NOT: llvm.intr.dbg.declare
-  llvm.intr.dbg.declare #di_local_variable = %1 : !llvm.ptr
+  llvm.intr.dbg.declare #di_local_variable #llvm.di_expr<[]> = %1 : !llvm.ptr
```
Comment (Mogball) [inline mlir/test/Dialect/LLVMIR/mem2reg-dbginfo.mlir:39]: suggestion: can you omit the exprs if they're empty?

File: mlir/test/Dialect/LLVMIR/debuginfo.mlir [hunk 2]
```diff
@@ -155,17 +155,17 @@ llvm.func @addr(%arg: i64) {
...
 }
 
 // CHECK: llvm.func @value(%[[ARG1:.*]]: i32, %[[ARG2:.*]]: i32)
 llvm.func @value(%arg1: i32, %arg2: i32) {
-  // CHECK: llvm.intr.dbg.value #[[VAR1]] = %[[ARG1]]
-  llvm.intr.dbg.value #var1 = %arg1 : i32
-  // CHECK: llvm.intr.dbg.value #[[VAR2]] = %[[ARG2]]
-  llvm.intr.dbg.value #var2 = %arg2 : i32
+  // CHECK: llvm.intr.dbg.value #[[VAR1]] #llvm.di_expr<[2, 4096, 0, 4]> = %[[ARG1]]
+  llvm.intr.dbg.value #var1 #llvm.di_expr<[2, 4096, 0, 4]> = %arg1 : i32
```
Comment (Mogball) [inline mlir/test/Dialect/LLVMIR/debuginfo.mlir:166]: question: is there any chance we can try to sugar the enums a bit?

File: mlir/test/Dialect/LLVMIR/mem2reg-dbginfo.mlir (hunk 1 above)
Comment (zyx-billy) [reply mlir/test/Dialect/LLVMIR/mem2reg-dbginfo.mlir:39]: oh sure I can try that

File: mlir/test/Dialect/LLVMIR/debuginfo.mlir (hunk 2 above)
Comment (zyx-billy) [reply mlir/test/Dialect/LLVMIR/debuginfo.mlir:166]: You mean printing the operator names? I added that as a TODO, mainly because it would involve refactoring some utils from llvm so that we can also understand the operators and how many arguments each has. I was planning on doing that as a followup if you don't have a strong preference.

Comment (Mogball) [reply mlir/test/Dialect/LLVMIR/debuginfo.mlir:166]: follow up is fine

File: mlir/test/Dialect/LLVMIR/debuginfo.mlir [hunk 3]
```diff
@@ -155,15 +155,15 @@ llvm.func @addr(%arg: i64) {
   %allocCount = llvm.mlir.constant(1 : i32) : i32
   %alloc = llvm.alloca %allocCount x i64 : (i32) -> !llvm.ptr
 
-  // CHECK: llvm.intr.dbg.declare #[[VAR0]] = %[[ALLOC]]
-  llvm.intr.dbg.declare #var0 = %alloc : !llvm.ptr
+  // CHECK: llvm.intr.dbg.value #[[VAR0]] = %[[ALLOC]]
+  llvm.intr.dbg.value #var0 = %alloc : !llvm.ptr
```
Comment (zero9178) [inline mlir/test/Dialect/LLVMIR/debuginfo.mlir:159]: Why is this change here necessary?

File: mlir/include/mlir/Dialect/LLVMIR/LLVMAttrDefs.td [hunk 4]
```diff
@@ -261,6 +261,17 @@ def LLVM_DITagParameter : LLVM_DIParameter<
...
 >;
 
+//===----------------------------------------------------------------------===//
+// DIExpressionAttr
+//===----------------------------------------------------------------------===//
+
+// TODO: Implement custom printer/parser for elements so that operators are
+// dumped in textual form.
+def LLVM_DIExpressionAttr : ArrayOfAttr<LLVM_Dialect, "DIExpression",
+                                          "di_expr", "uint64_t"> {
```
Comment (Mogball) [inline mlir/include/mlir/Dialect/LLVMIR/LLVMAttrDefs.td:271]:
```suggestion
                                        "di_expr", "uint64_t"> {
```

style nit

**Review by Mogball — APPROVED** [review]

File: mlir/test/Dialect/LLVMIR/debuginfo.mlir (hunk 3 above)
Comment (zyx-billy) [reply mlir/test/Dialect/LLVMIR/debuginfo.mlir:159]: oops, I might've accidentally changed this. Will revert, thanks!

**Review by zero9178 — APPROVED** [review]
LGTM

**Review by gysit — APPROVED** [review]
Thanks for adding DIExpression support.

LGTM and looking forward to the follow up :)
